panels/models/attraction.py

```diff
@@ -187,13 +187,51 @@ def groupify(items, keys, val_key=None):
     return groupified
 
 
+def mask(s, mask_char='*', min_unmask=1, max_unmask=2):
+    s_len = len(s)
+    if s_len <= min_unmask:
+        return s
+    elif s_len <= (2 * max_unmask):
+        unmask = max(min_unmask, math.ceil(s_len / 2) - 1)
+        return s[:unmask] + (mask_char * (s_len - unmask))
+    return s[:max_unmask] + (mask_char * (s_len - max_unmask))
+
+
+RE_NONDIGIT = re.compile(r'\D+')
+RE_SLUG = re.compile(r'[\W_]+')
+
+
+def sluggify(s):
+    return RE_SLUG.sub('-', s).lower().strip('-')
+
+
 @Session.model_mixin
 class Attendee:
+    NOTIFICATION_EMAIL = 0
+    NOTIFICATION_TEXT = 1
+    NOTIFICATION_NONE = 2
+    NOTIFICATION_PREF_OPTS = [
+        (NOTIFICATION_EMAIL, 'Email'),
+        (NOTIFICATION_TEXT, 'Text'),
+        (NOTIFICATION_NONE, 'None')]
```

Comment:
In short, it's because I'm lazy and most of this will be redone next year.

Going into more detail, I have a real love/hate relationship with `[enums]` (much like I do with `@suffix_property`). I have a much larger rant about `[enums]`, but instead of wailing and gnashing my teeth, I want to talk a little bit about the distinction between code, constants, and configuration.

Presumably, you'd like `NOTIFICATION_PREFS` to exist in the _config_ because you'd like it to be _configurable_, right? Not every deployment will have a twilio account (or will otherwise be able to send text messages). So you'd like to _configure_ that on a per deployment basis, right?

Here's the thing though. In this implementation, I explicitly reference the individual enum values: `NOTIFICATION_EMAIL`, `NOTIFICATION_TEXT`, and `NOTIFICATION_NONE`. So, lets pretend we've implemented this in an `[enums]` section of the config. What happens when we add a new value to the enum?

Let's pretend we add `NOTIFICATION_PIGEON`, because we want to be able to notify attendees using carrier pigeon. What happens? Well, nothing happens. The code is not aware of the existence of `NOTIFICATION_PIGEON`. It's behavior will not change based on the addition of that value.

Instead of _adding_ a value to the enum, let's consider what happens when we _remove_ a value from the enum. Consider a deployment that does not have a twilio account, and cannot send text messages. We'd like to entirely disable text messages. What happens when we remove `NOTIFICATION_TEXT` from the enum?

Everything breaks, of course. References to `NOTIFICATION_TEXT` are hardcoded in the source code. As soon as those lines of code are executed, the program will fail catastrophically. Due to the way this feature has been written, we _cannot_ remove `NOTIFICATION_TEXT`. The code relies the existence of that value.

To put it another way, it is not _configurable_. It is not a thing we can configure. It does not belong in the configuration. It is a constant that belongs in the code.

Now, I know what you're thinking. "But we _do_ want to configure this! This is precisely the kind of thing that we _want_ to be able to configure!" And I agree! However, we should not delude ourselves into believing that storing these values in the configuration immediately makes them configurable. It does not! If we put these values into the config – without changing our implementation at all – then all we've accomplished is to mix our source code with our configuration.

You may as well be asking why the various [socket options](https://github.com/freebsd/freebsd/blob/master/sys/sys/socket.h#L102) are declared in source code instead of in a configuration file.

It's because the socket implementation has an inherent understanding of each of those constants. Those constants have an explicit meaning to the source code. The source code expects those constants to exists, and it would break if they didn't.

A good rule of thumb to follow when deciding whether or not to use `[enums]` is to ask yourself if the code will break when a value is removed. This implementation will certainly break, because it expects each of those values to exist, so I kept it out of the configuration.

# There has to be a Better Way<sup>TM</sup>

But back to the original point, we definitely _want_ the available notification options to be configurable. So what's the solution?

This is where we start talking about dependency injection. What we really want to do is create a well-defined interface, like `NotificationProvider` or something. And then in our config, we can add or remove `NotificationProviders` which _will_ change the behavior of the source code. We don't need to know about the internals of a `NotificationProvider`, only what the interface defines.

A good example of this that we're all familiar with is the python logging system. You configure log handlers by adding them in your config file. You don't need to understand the internals of a log handler. Some of them write to a file. Some of them send data over the network. You might even have a log handler that emails you. Someone in the future might write a log handler that we couldn't possibly imagine, and it will still work without modifying the current codebase.

So what you're really talking about is being able to configure our notification preferences the way we configure our logging. And that's what I'd love to do if we had more time. But to quote [Devops Rumsfeld](https://twitter.com/agentdero/status/200372798424760320):
> You go to production with the code you have, not the code you wish you had

I'm going to take it a step further, though. I'm going to say that even in a perfect world where we had the time to do this feature right, we still wouldn't want to put it in the config file. If you step back and think about it, we are not configuring the _system_ as a whole. We are configuring a single _event's_ ability to send notifications. Perhaps Super MAGFest has the budget to support text messaging, but MAGStock does not. Or maybe text messaging just doesn't make sense for MAGStock because we're in the woods and there's no cellular reception (_I actually don't think that's true_).

In any case, we'd want to be able to configure notification options _per event_ not _per system_. In which case, we'd want to store notification options in the _database_, not the configuration file. We might still keep a list of _available_ `NotificationProviders` in the config. But even then, I'd argue we'd be better off importing them dynamically, perhaps from plugins, and using introspection to discover what `NotificationProviders` are available.

Reply:
Fair enough!

FWIW, ``[enums]`` came about for somewhat different reasons that you seem to think.  To be honest, if I'd known as much about metaclasses then as I do now then I would have done things a lot differently.

What I noticed was that I had a lot of code in the old days that looked like this:

```python
SMALL = 1
MEDIUM = 2
LARGE = 3
SIZE_OPTS = (
    (SMALL, 'small'),
    (MEDIUM, 'medium'),
    (LARGE, 'large')
)
SIZES = dict(SIZE_OPTS)
```

So what's wrong with that?  It's not fundamentally bad code or anything, but it's twice as long as it needs to be.  I almost never care about the actual integer values of my enum options, but I still end up having to devote two lines of code instead of one to declaring them.  I also wanted to be able to automatically have both an ordered list/tuple version and a dictionary option without having to add a line of code every single time.

In other words, I wanted a declarative way to generate this kind of config.  If I had to do it over again, I would have used a metaclass to do something like

```
class Sizes(ConfigEnum):
    SMALL = 'small'
    MEDIUM = 'medium'
    LARGE = 'large'
```

With the right metaclass, I could have done almost anything I wanted, including:
- Automatically generating integer constants for each field.
- Automatically generating fields to meet our current use cases of having things like ``c.SIZE_OPTS``, ``c.SIZES``, and ``c.SIZE_VARS``
- Overriding the defined-in-code values in a config file if I wanted.  For example, suppose we wanted to change ``'medium'`` to ``'middle'``.  We could have a special ``[enums]`` subsection called ``[[sizes]]`` in our config file with the ability to override the default text, which would still give us the "just edit a config file instead of having to edit the code to change a label" feature.

Even before I knew about metaclasses, the main thing which stopped me from trying to hack something together using this approach was ordering.  We usually want our ``_OPTS`` to be in the order in which they were declared.  There are a couple of ways we could have done that, e.g.

1) Use introspection of the source code to figure out what line the value was declared on.  That's pretty janky and magicky (even for me!), so I didn't want to do anything like that.

2) Instead of using raw strings, wrap your strings in some kind of function call which can increment a counter behind the scenes, e.g.

```python
class Sizes(ConfigEnum):
    small = val('small')
    medium = val('medium')
    large = val('large')
```

but I really hated how that looked.

Thankfully now we have the ``enum`` module, which preserves ordering, so if ``Sizes`` was an enum subclass then we would know the order in those options were declared.  (Enums already have a metaclass they use for a lot of their functionality, and I'm not sure how easy it is to hook into that, but I think it would still give me a way to solve the most annoying part of my use case.)

Anyway, the TLDR of all of this is that ``[enums]`` was really less about "everything should just be configurable" and more about "Eli hates code duplication and wants a declarative way to define these kinds of options".

There are obviously some intended benefits like being able to change the textual label for an enum field.  I also liked being able to define a base set of things like badge types and then add new options onto that in config, while allowing the configspec to mandate that the "everything breaks if you don't have this" options are still there.

I get what you're saying though.  While I still like my overall approach to ``[enums]`` from a "good declarative way to get what we need" perspective, it creates a tight coupling of code and config which is pretty bad from a maintainability perspective.  It's one of a number of design decisions that probably stopped being correct as soon as we started actually using these config files as, like, actual config files.

(Though as a philosophical issue, I generally think it's better to think of a ``configspec.ini`` as being *code* whereas a ``development-defaults.ini`` is obviously *config*.  I realize not everyone agrees with this, since a configspec is by definition a config file, but I think it's more accurate to think of it as code.  Not that this matters much to your overall point.)

Regardless, you'll certainly have my blessing to tear up the ``[enums]`` code whenever you get to a point where it makes sense to do that.  I'll be disappointed if we end up with a two-lines-of-code-instead-of-one but that's not a big deal in the grand scheme of things.

Reply:
If we're just trying to cut down on boilerplate, I think we could solve that problem with some pretty straightforward code (ignoring the fact that [python now has enum](https://docs.python.org/3/library/enum.html)). No magic involved:

```python
>>> class enum:
...     def __init__(self, *args, **kwargs):
...         self._VARS = args
...         self._OPTS = []
...         count = 0
...         for arg in args:
...             setattr(self, arg, count)
...             self._OPTS.append((count, kwargs.get(arg, arg.title())))
...             count += 1
...         self._VALS = dict(self._OPTS)
... 
>>> class Attendee:
...     NOTIFICATION_PREFS = enum('NONE', 'EMAIL', 'TEXT', TEXT='Text/SMS')
... 
>>> Attendee.NOTIFICATION_PREFS._VARS
('NONE', 'EMAIL', 'TEXT')
>>> 
>>> Attendee.NOTIFICATION_PREFS._OPTS
[(0, 'None'), (1, 'Email'), (2, 'Text/SMS')]
>>> 
>>> Attendee.NOTIFICATION_PREFS._VALS
{0: 'None', 1: 'Email', 2: 'Text/SMS'}
>>> 
>>> Attendee.NOTIFICATION_PREFS.NONE
0
>>> Attendee.NOTIFICATION_PREFS.EMAIL
1
>>> Attendee.NOTIFICATION_PREFS.TEXT
2
```

Reply:
I definitely considered something like that, and I did in fact have an ``enum`` class that was similar to that many years ago.  However, in the (fairly common) case where you want a custom label for all of your options, your implentation still requires us to write each option twice, once for ordering and once for labeling.  (Thankfully, Python has ``enum`` now which handles the ordering for us.)

In general I think we're in agreement though, since I think we both have roughly the same opinion that ``[enums]`` is worse than this general approach of having a declarative class.

Reply:
OMG, you're killin' me brah!

```python
>>> import six
>>> class enum:
...     def __init__(self, *args):
...         self._VARS = []
...         self._OPTS = []
...         count = 0
...         for arg in args:
...             if isinstance(arg, six.string_types):
...                 label = arg.replace('_', ' ').title()
...             else:
...                 label = arg[1]
...                 arg = arg[0]
...             setattr(self, arg, count)
...             self._VARS.append(arg)
...             self._OPTS.append((count, label))
...             count += 1
...         self._LABELS = dict(self._OPTS)
... 
>>> NOTIFICATION_PREFS = enum('NONE', 'EMAIL', 'CARRIER_PIGEON', ('TEXT', 'Text/SMS'))
>>> 
>>> NOTIFICATION_PREFS._VARS
['NONE', 'EMAIL', 'CARRIER_PIGEON', 'TEXT']
>>> 
>>> NOTIFICATION_PREFS._OPTS
[(0, 'None'), (1, 'Email'), (2, 'Carrier Pigeon'), (3, 'Text/SMS')]
>>> 
>>> NOTIFICATION_PREFS._LABELS
{0: 'None', 1: 'Email', 2: 'Carrier Pigeon', 3: 'Text/SMS'}
>>> 
>>> NOTIFICATION_PREFS.NONE
0
>>> NOTIFICATION_PREFS.EMAIL
1
>>> NOTIFICATION_PREFS.CARRIER_PIGEON
2
>>> NOTIFICATION_PREFS.TEXT
3
```

Reply:
Haha, I was honestly thinking of trolling you by saying "You forgot to do a ``.replace('_', ' ')`` for the description :P" and then I say that you literally did make that improvement between your last version and this one, along with Python 2/3 compatibility for some reason :rofl: 

I do really like this approach though.  I would totally support doing something like this.  (I'd probably replace the autoincrementing ``count`` with the hash-based approach we're using for ``[enums]`` though, since otherwise you wouldn't be able to insert new items in the middle without changing the values of all of the existing items which come after that.)

I will say that I was sad that one of the earlier ``enum`` proposals got dropped, which would have allowed us to specify enum options without assignment, e.g.

```python
class NotificationPrefs(ConfigEnum):
    NONE
    CARRIER_PIDGEON
    TEXT = 'Text/SMS'
```

but they eventually decided that violated Python's distinction between expressions and statements.  Which is fair, but that's pretty much the only thing I could ask for that would make this better than your overall approach to ``enum`` :)